include/fiction/algorithms/physical_design/post_layout_optimization.hpp

```diff
@@ -339,6 +339,71 @@ void optimize_output_positions(Lyt& lyt) noexcept
         }
     }
 }
+
+/**
+ * Utility function that checks and optimizes PO positions after each gate relocation iteration.
+ * This function moves POs that are not optimally positioned (e.g., in second rightmost or second bottom positions)
+ * to the optimal border positions by inserting buffer gates where the POs were.
+ *
+ * @tparam Lyt Cartesian gate-level layout type.
+ * @param lyt Gate-level layout.
+ * @param moved_gates Moved gates counter to decrement if PO is moved.
+ */
+template <typename Lyt>
+void check_and_optimize_po_positions(Lyt& lyt, uint64_t& moved_gates) noexcept
+{
+    // check all POs are in right (x=lyt.x()) or bottom (y=lyt.y()) border
+    lyt.foreach_po(
+        [&lyt, &moved_gates](const auto& po) noexcept
+        {
+            if (const auto tile = lyt.get_tile(lyt.get_node(po));
+                !(lyt.is_at_eastern_border(tile) || lyt.is_at_southern_border(tile)))
+            {
+                if ((tile.x == lyt.x() - 1) && (lyt.is_empty_tile({lyt.x(), tile.y})))
+                {
+                    // get fanin signal of the PO
+                    std::vector<mockturtle::signal<Lyt>> signals{};
+                    signals.reserve(1);
+                    lyt.foreach_fanin(lyt.get_node(tile), [&signals](const auto& fanin) { signals.push_back(fanin); });
+
+                    // move PO to the rightmost border
+                    lyt.move_node(lyt.get_node(tile), {lyt.x(), tile.y, 0}, {});
+
+                    // create a buffer at the previous location of the PO and connect it with its fanin
+                    lyt.create_buf(signals[0], {lyt.x() - 1, tile.y});
+
+                    // connect the PO with the new buffer
+                    lyt.move_node(lyt.get_node({lyt.x(), tile.y}), {lyt.x(), tile.y, 0},
+                                  {lyt.make_signal(lyt.get_node({lyt.x() - 1, tile.y}))});
+
+                    --moved_gates;
+                }
+                else if ((tile.y == lyt.y() - 1) && (lyt.is_empty_tile({tile.x, lyt.y()})))
+                {
+                    // get fanin signal of the PO
+                    std::vector<mockturtle::signal<Lyt>> signals{};
+                    signals.reserve(1);
```

Comment:
`.reserve(1)` is most likely superfluous if you plan to store only a single element anyway.